File: test_programs/memory.py
#!/usr/bin/env python3
"""
Memory allocation test program to verify memory limits
"""
import sys
import time
import gc

def test_memory_allocation():
    print("=== Memory Allocation Test ===")
    
    # Get initial memory info
    try:
        import psutil
        process = psutil.Process()
        initial_memory = process.memory_info().rss / 1024 / 1024
        print(f"Initial memory usage: {initial_memory:.2f} MB")
    except ImportError:
        print("psutil not available, using basic memory test")
    
    # Test small allocation
    try:
        small_data = bytearray(1024 * 1024)  # 1 MB
        print("✓ Small allocation (1 MB) successful")
        del small_data
        gc.collect()
    except MemoryError:
        print("✗ Small allocation failed")
        return
    
    # Test medium allocation
    try:
        medium_data = bytearray(10 * 1024 * 1024)  # 10 MB
        print("✓ Medium allocation (10 MB) successful")
        del medium_data
        gc.collect()
    except MemoryError:
        print("✓ Medium allocation blocked by memory limit")
    
    # Test large allocation
    try:
        large_data = bytearray(100 * 1024 * 1024)  # 100 MB
        print("⚠ Large allocation (100 MB) successful")
        del large_data
        gc.collect()
    except MemoryError:
        print("✓ Large allocation blocked by memory limit")
    
    # Test progressive allocation to find limit
    allocations = []
    chunk_size = 5 * 1024 * 1024  # 5 MB chunks
    total_allocated = 0
    
    print(f"Testing progressive allocation in {chunk_size // 1024 // 1024} MB chunks...")
    
    for i in range(50):  # Try up to 250 MB
        try:
            chunk = bytearray(chunk_size)
            allocations.append(chunk)
            total_allocated += chunk_size
            print(f"✓ Allocated chunk {i+1}: {total_allocated // 1024 // 1024} MB total")
            time.sleep(0.1)  # Brief pause
        except MemoryError:
            print(f"✓ Memory limit reached at {total_allocated // 1024 // 1024} MB")
            break
    
    # Clean up
    del allocations
    gc.collect()
    print("✓ Memory cleanup completed")

if __name__ == "__main__":
    test_memory_allocation()

File: rustbox/test_programs/memory.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define MB (1024 * 1024)

int main() {
    printf("=== C Memory Allocation Test ===\n");
    
    void *ptrs[200];  // Array to hold pointers
    int allocated_chunks = 0;
    size_t chunk_size = 5 * MB;  // 5 MB chunks
    size_t total_allocated = 0;
    
    // Test small allocation first
    void *small = malloc(MB);
    if (small) {
        printf("✓ Small allocation (1 MB) successful\n");
        memset(small, 0xAA, MB);  // Touch the memory
        free(small);
    } else {
        printf("✗ Small allocation failed\n");
        return 1;
    }
    
    // Test progressive allocation to find memory limit
    printf("Testing progressive allocation in %zu MB chunks...\n", chunk_size / MB);
    
    for (int i = 0; i < 200; i++) {
        ptrs[i] = malloc(chunk_size);
        if (ptrs[i] == NULL) {
            printf("✓ Memory allocation failed at chunk %d (limit reached)\n", i);
            printf("✓ Total allocated before failure: %zu MB\n", total_allocated / MB);
            break;
        }
        
        // Touch the allocated memory to ensure it's actually available
        memset(ptrs[i], 0x55, chunk_size);
        
        allocated_chunks++;
        total_allocated += chunk_size;
        
        printf("✓ Allocated chunk %d: %zu MB total\n", i + 1, total_allocated / MB);
        
        // Add a small delay
        usleep(100000);  // 100ms
        
        // Stop if we've allocated too much (safety check)
        if (total_allocated > 500 * MB) {
            printf("⚠ Allocated over 500 MB, stopping test\n");
            break;
        }
    }
    
    printf("Final stats: %d chunks allocated, %zu MB total\n", 
           allocated_chunks, total_allocated / MB);
    
    // Clean up allocated memory
    for (int i = 0; i < allocated_chunks; i++) {
        if (ptrs[i]) {
            free(ptrs[i]);
        }
    }
    
    printf("✓ Memory cleanup completed\n");
    
    return 0;
}